File: src/0096-Unique-Binary-Search-Trees/0096.py
class Solution:
    def numTrees(self, n):
        """
        :type n: int
        :rtype: int
        """
        mem = [0]*(n+1)
        mem[0], mem[1] = 1, 1
        
        for i in range(2, n + 1):
            for j in range(1, i + 1):
                mem[i] += mem[j - 1]*mem[i - j]
                
        return mem[-1]

if __name__ == "__main__":
    n = 4
    print(Solution().numTrees(n))

File: src/0096-Unique-Binary-Search-Trees/0096.cpp
#include <iostream>
#include <string>
#include <vector>
#include <stack>
using namespace std;

static int x = []() {std::ios::sync_with_stdio(false); cin.tie(0); return 0; }();
class Solution 
{
public:
    int numTrees(int n) 
    {
        vector<int> mem(n + 1, 0);
        mem[0] = 1, mem[1] = 1;
        for (int i = 2; i <= n; ++i)
        {
            for (int j = 1; j <= i; ++j)
            {
                mem[i] += mem[j - 1]*mem[i - j];
            }
        }
        return *rbegin(mem);
    }
};
int main()
{
    return 0;
}